### ISSUE TITLE
Think through behavior of frame-timing wrt :visited changes

### ISSUE THREAD

**Issue by natduca** [WICG/frame-timing#40]
From Jonas: "Please make sure that this doesn't expose things like paints that
happen due to, for example, a link changing color due matching a :visited rule."

We should think through how FT works in this case.

**Comment by igrigorik** [comment]
Link to thread: https://groups.google.com/d/msg/mozilla.dev.platform/5dzctcjEhSw/rCaj4NEoRHwJ

**Comment by igrigorik** [comment]
As of today, and unless we add some mitigations for it, it would be possible (I think) to use FT to observe :visited changes. The simple example is...
1. Create a document with a link that points to a bogus url (not visited) and some fixed text.
2. Set a timeout that's >16ms (or use raF) and within it change the href to a URL you want to probe.
3. Wait and observe if a PerformanceRenderTiming event is fired after 16ms - i.e. if we paint.

The 16ms interval places an upper bound on how many links you can enumerate through per second, but that's likely not sufficient. Reviewing related history in [1], [2]: we're already telling "white lies" and placing limits on what you can do with :visited... seems like we should do something similar to account for this case as well.

After talking to a few folks, one idea: disable `:visited` entirely while/FT is active? Any thoughts or suggestions on this, or other strategies we should consider?

[1] https://developer.mozilla.org/en-US/docs/Web/CSS/Privacy_and_the_:visited_selector
[2] http://dbaron.org/mozilla/visited-privacy

/cc @sicking @dbaron @mpb @rbyers

**Comment by hiikezoe** [comment]
Please let me clarify about disabling :visited.

a) Do not render :visited change while observing FT.
b) Do not include render timing entry of :visited change while observing FT.
Which one did you mean?

**Comment by mpb** [comment]
a) I believe.

**Comment by igrigorik** [comment]
@hiikezoe (a). I'm not sure that (b) is even feasible.. since by the time this stuff gets plumbed down the stack, we may no longer know which pixel change was triggered by what property.

**Comment by bzbarsky** [comment]
@igrigorik Seems to me like the right mitigation here is that when the href of the anchor changes the anchor gets repainted, without regard to whether its visited state changed.  You need something like that even without FT, because otherwise you can still use the repaint happening or not to perform timing attacks, I'd think.

**Comment by bzbarsky** [comment]
A bigger issue is that if you just start with an anchor with some href there is probably still an extra paint in the visited case because the is-this-visited query is async...

**Comment by paulirish** [comment]
In Chrome's case we know the reason of our invalidations leading to a paint or composite event. So we know a :visited link changing style contributed to this event.  Can you confirm  @progers ?

**Comment by progers** [comment]
We could mark a bit if a :visited link changed in any given frame but we don't have it at the moment.

An approach that may work is to just spec performance monitoring as painting as if the :visited pseudo class is not applied. It's a strange edge-case but I can't think of another approach. We already have an expensive color check for links so this isn't onerous in Blink.
Edit: this is (a), above.

**Comment by igrigorik** [comment]
FWIW, a test case for current behavior: http://output.jsbin.com/wogelo/13/quiet
- **IE, Safari:** render right color based on :visited state of original `href` value, but then ignore `href` attribute updates - i.e. the original color persists after it is initially painted.
- **Firefox, Chrome:** both repaint link color on `href` attribute updates.

As such, IE and Safari behavior would not trigger any additional paint events through Frame Timing, and -- I think -- everything continues to work as is without any additional privacy implications. That said, there is one _existing_ gotcha with this behavior: if you're recycling a DOM subtree (e.g. a post snippet with a link) and changing links.. the visited state persists. This is not great, but presumably this is already the case on Webkit and IE, so this is nothing new.

A few other alternatives:
1. Disable :visited when FT is enabled. This is similar to existing IE/Safari behavior, but would only apply when FT is on... That said, I would expect that many apps enable FT as a continuos observer, at which point we're back to identical behavior with IE/Safari -- modulo first link paint, which is async in Chrome/FF and should probably still happen? 
2. Repaint anchor when `href` attribute is touched regardless of :visited - i.e. repaint same color if you have to. This provides nice(r) UX properties and addresses the recycling DOM fragment use case.

@toddreifsteck curious to hear your thoughts from IE's perspective. How did you guys arrive at current behavior, any pros/cons to it vs (2)?

**Comment by esprehn** [comment]
We can't degrade the user experience when the page is running the timeline. This isn't just about href changing though, if you have two tabs open and I click a link in one tab we need to repaint the link in all open tabs. Using that I can observe what pages you have visited.

I think the only way to make this work is to repaint every link the page every time we try to update the visited state of any link.

**Comment by esprehn** [comment]
Specifically a feature for debugging and reporting metrics should not make the UX worse, so we can't change the behavior of :visited for users with this running.

**Comment by toddreifsteck** [comment]
Adding @DigiTec to this issue

**Comment by DigiTec** [comment]
@esprehn Hum, I'd like to dig in a bit more on the idea that updating the href in one tab does/should apply to all other tabs. In a multi-process architecture it is non-trivial to maintain a "link dirty" state for "other tabs" that are running. In fact, you might have tens or even hundreds of tabs open and so immediate propagation (and therefore paging in of the memory for that tab instance) or deferred propagation (maintaining a flag) are very cost prohibitive. Also, re-processing all links on a page based on the dirty state when say visibility changes swap you in is also very cost prohibitive.

Given this, and speaking psuedo-definitively on the IE and Edge engines, I don't believe we try to update other tabs at all when a link changes in one tab. In fact, I don't think we update them within the same tab either. I notice that Chrome does update links through some basic tests I cooked up and so arguably there would be a user experience reduction in some cases if all browsers chose to adopt the same approach.

But the reality is, we've already scoped the kinds of updates that are allowed to so few properties, that you have to question how much user experience impact there really is.

I'll try to reply to @igrigorik regarding the alternatives proposed above. There are a few questions there that require some investigation on my part. For instance, around the fragment re-use case.

**Comment by esprehn** [comment]
Chrome, Safari and Firefox all update so it's Edge/IE that's the odd one
out here. This is not something we're going to change, visited links are an
important feature of the web.

Note that the update is not very expensive, it's a shared memory segment
between the processes and we notify the other pages async. Users can't see
hundreds of tabs at the same time, you only need to update visible ones.

I'd say it's a pretty bad Edge bug if visited links in the same page don't
even update.

**Comment by DigiTec** [comment]
@esprehn You still need a map somewhere of which anchors to invalidate based on their hrefs, and given that an average page has more than 150 such external links (this is old telemetry, take it with a grain of salt). Additional, at least for Edge, being a platform component, we can have more things interacting with the cache containers for history than just the browser frame. There can also be Windows Web Applications, Third-party applications, etc... 

We also agree visited is useful, especially within the same page, so I don't think we are arguing visited is useful or not, however, visited comprises multiple use cases and we can probably demonstrate the relative user value of some of them as not being as important as others. I think that is the point of this discussion in fact, is to discuss what we "could do". Many sites (I think we had data that showed more than half of the top sites when we did the first link coloring mitigation around IE 8/9) color them identically and so they are really more of an author feature and not so much a user feature. Exceptions being users who override author styles and colors with their own, which I also think is a very, very low percentage.

**Comment by DigiTec** [comment]
@igrigorik Okay, so more data on our end. Our async behavior for updating links is gated on whether or not we performed the async check. We only perform the async check on an element once, which is why, when you update the `href` property you don't get further invalidation. We reset the async check flag whenever an element is involved in a tree splicing mutation. So for instance, if you are cloned, the flag is reset. If you are moved from point A to point B the flag is reset.

This means most templating functionality would work. However, databinding type technologies would not work. If they databind and only change specific attributes or properties and don't remove and reinsert the element into the tree, then our async check flag won't get reset and the link won't be repainted.

For (2) If we always repaint anchors, then wouldn't we have to also say that during initial page load, we'd have to invalidate all anchors once our async workload for visited detection returns regardless of whether or not the links are visited or not? In order to ensure that the `render` duration does not provide a detectable difference? For large pages this could be a prohibitive. Maybe there is something that can allow us to avoid this, but I can't immediately think of one.

Maybe it would be useful to build a corpus of test pages with desired behaviors and render timing outputs as well. We could probably clarify some of the render timings documents with a frame callback and timing diagram of some sort. I think some of this is in the explainer document, but it would be useful to tweak the visuals to match the spec language and structures.

@esprehn The more I think about these timing and invalidation items, the more interesting the privacy concerns are for dynamically updating "other tabs" when one tab updates the browser history. Much care would have to be taken to ensure that when the tab is in the background and a new history entry is added, that you NEVER update that tab. But Chrome has no concept of this. If I have two frames open, and I click a link in one and I happen to have that link in another, I'll get a frame timing. A malicious site could stand on the side and just wait for this to happen. Worse, a malicious ad could sit in its iframe with thousands of links, waiting to see which ones you might click on within the same session.

Open the following in two tabs that are in separate frames (this is to make it visibly obvious, but you could do iframes, etc...)
http://jsfiddle.net/z55dmoyq/
1. Click the link on one
2. Note the color change in the other.

**Comment by igrigorik** [comment]
@esprehn @DigiTec fwiw, I don't want to put a damper on it, but the cross-process discussion is starting to feel like a separate thread. That said, not sure I have a particularly good suggestion for where to move it either.

> For (2) If we always repaint anchors, then wouldn't we have to also say that during initial page load, we'd have to invalidate all anchors once our async workload for visited detection returns regardless of whether or not the links are visited or not? In order to ensure that the render duration does not provide a detectable difference? For large pages this could be a prohibitive.

Presumably this is already the case for any page where the user has visited many/most links, no? As in, you'd have to repaint all of those links once the async query gets back. The difference here is that we're asking UA to repaint the link when the query returns.. regardless of its current color.

**Comment by DigiTec** [comment]
@igrigorik In general, there are hundreds of links on a page only a few of which are actually visited. So we tend to send off a package of 150-300 URL's to be considered. Of those, only fewer than 10 are invalidated. So if we had to invalidate all 300, that would probably be a blip on the radar from a performance perspective. I think this would be a concern.

**Comment by igrigorik** [comment]
@DigiTec doesn't that acknowledge that current behavior is an attack vector? If I create a page with a few thousand anchors, all pointing to the same page... if it's a blip on the radar than I know you've visited that page (i.e. by forcing an expensive repaint). Forcing a repaint would introduce a perf cost, but it does mitigate (I think) this particular timing attack...

**Comment by DigiTec** [comment]
@igrigorik I haven't entirely ruled out that there isn't an existing attack vector, there may well be. Timing attacks are becoming increasingly more sophisticated and timing measures, with high performance timings, have made it easier and easier to measure small amounts of time and therefore use basic scaling to measure quite accurately relative time differences.

In IE we may do multiple link highlighting packages, but I don't see that we put a cap that would "force" one to get kicked off and thus break up the timing. I'll look into that more deeply though. When you request an animation frame callback today, you get one regardless of whether or not there is a paint. So you basically have to detect that paints got held up for some amount of time. Perhaps breaking it up and minimizing the link highlighting work per frame would be useful. It would potentially be a bad user experience as well.

**Comment by igrigorik** [comment]
Aside, but relevant.. With remove-clone-insert technique, all browsers repaint link color: http://output.jsbin.com/wogelo/quiet

**Comment by igrigorik** [comment]
I believe this is addressed via #48. See: https://w3c.github.io/frame-timing/#privacy-and-security.

Closing, but please feel free to reopen if you feel otherwise.
